### DISCUSSION TITLE
How do I change the location of files to download?

### DISCUSSION THREAD

**Discussion by Tanggue** [aclap-dev/video-downloadhelper#744] [Help and Q&A]
### Your question

Due to this newly updated design, I am unable to find a way to change the download location.
For example, I want to change the download location from C drive to D drive. Where can I change it?

version: 9.0.1.13 (stable), mozilla

### Addon & CoApp version and details

9.0.1.13 (stable), mozilla

**Comment by GrampaWildWilly** [comment]
That is a feature you can take advantage of only if you use the CoApp.

**Reply by Awesome820** [reply]
whats the CoApp?

**Reply by Tanggue** [reply]
In the previous version, you could change it right from the settings button. Do I need CoApp in the new version?
What is CoApp?

**Reply by GrampaWildWilly** [reply]
I do not work for VDH.  I'm just another user.

Wow . . .  Where have you guys been?  Back in the old days of Firefox before version 57 (the new Firefox released earlier today is 126.0.1, so we're talking like prehistoric times), browser plugins (their old name) used to be able to do a lot of things on your system.  One of those things was writing files into directories.  But the rules changed with Firefox 57, also called Quantum.  Beginning at that moment, browser add-ons or extensions (new terminology at that time) were no longer permitted to write files on the user's HDDs.  The browser itself still could, but add-ons could no longer.  This was considered an improvement in software security.  Maybe so.  Seems a bit harsh to me.  But that's how they defined the new world.

Now think about it.  VDH downloads videos.  And where do they go?  And how can VDH put them there if it's not permitted to write to your HDDs?  See the problem?  To circumvent this new rule, Michel came up with the idea of the Companion Application, affectionately nicknamed the CoApp.  At a certain moment during VDH's processing of a video, it launches execution of the CoApp in its own address space, separate from the browser.  The CoApp then does the actual reading of data from the Internet & writing it on your HDDs.  It's not within the control & supervision of Firefox so it can do whatever it wants, including writing files into directories on your HDDs.  It's also not an application like most others on your system.  You can't execute it.  Only VDH can do that.

At the same time Michel invented the CoApp, he had to completely rewrite VDH.  That was the transition from VDH 6 to VDH 7.  It is not a recent development.  We're on VDH 9 these days.  That's why I'm somewhat shocked you're asking your questions at this late date.

I've been talking about just Firefox.  I'm not entirely clear if Firefox was the first browser to make this security change for add-ons.  I'm not entirely clear whether Firefox, Chrome, & Edge all co-operated or at least exchanged ideas & agreed on protocols.  But all the browsers have adopted this not-so-new-now new rule.  Although there are 3 separate products -- VDH for Firefox, VDH for Chrome, VDH for Edge -- there is only one CoApp & it works in all environments.  It's an application that runs separate from the browser so it makes sense that there is only one CoApp & all 3 versions of VDH can share it.  But when I say one CoApp, I mean one CoApp on your system.  Of course, the CoApp is different between Windows & Mac & Linux.

At the top of every page of this forum you will find a link to the Wiki.  The Wiki is more or less our user manual.  You would do well to read that.  There is some discussion in there of the CoApp.  But what I explain above is not in the Wiki.  It should be, but it's not.  That's why I wrote that.  I shouldn't have had to, but I did.  On the other hand, the Wiki is where you will find the CoApp, along with instructions for how to install it.

VDH exists because there are many types of media that the browsers can't download themselves.  They can play lots of them.  But they can't necessarily download them.  That was the original impetus behind VDH many years before I ever discovered the product.  Today, the CoApp provides the ability to download HLS & DASH streams as well as the queer duck known as YouTube.  If you are presented with a simple file, like say an MP4, VDH offers the ability to download that using the browser itself, because that's a kind of file the browser could download without VDH's help.  But most of the streams available around the web these days are beyond the ability of the browsers to deal with.  That's where VDH has to use the CoApp.

As for the old setting for the download directory, that is still in the new VDH 9 settings.  Click the settings gear, which is now located at the bottom LEFT of the main VDH menu, no longer the right.  That gives you a list of a few settings, ending with an entry for More Settings.  Go look at the More Settings.  It's a scrollable list.  Scroll it.  You'll find the download directory.  It works much like it always used to.

But it's so much more convenient to use the CoApp.  You should not feel any trepidation about it.  Yeah, yeah, you'll get a watermark most of the time if you use the CoApp.  So?  Buy a license.  Don't be a cheapskate.  The license is not a subscription.  It is for life.  Your life, not your computer's.  You can use the same license on every computer you own now & every computer you will ever own for as long as you breathe.  That makes the license cheaper by the day the longer you use it.  But there are 3 separate VDH licenses, one for VDH for Firefox, a different one for VDH for Chrome, a different one for VDH for Edge.  The licenses are available on Windows & Mac.  The same license for, let's say Firefox, works on both Windows & Mac.  Linux is a special case.  Michel used to develop VDH on Linux so he has a soft spot in his heart for Linux users.  No license is needed on Linux.  Paul has pretty much taken over development of VDH since about a year ago.  I'm not entirely sure what platform he uses for development.  I have a feeling he has all 3 platforms at his disposal & he uses them all.

And due to a legal restriction imposed by Google, the owner of both Chrome & YouTube, VDH for Chrome does not download from YouTube.  Google has no power over Firefox nor Edge, so those versions of VDH download YouTube videos without any problem.

Aren't you glad you asked?

**Comment by gmeades** [comment]
Version 9.0.2.3 (CoApp version 2.0.19*)

Note: The default Video Downloadhelper download directory will be the C:\user\<current user>\dwhelper directory 

I had the same problem of not being able to change the Video Downloadhelper download directory to the one I wanted after a recent update, but finally figured it out.  Here are the steps I took - 

1. Click on the Manage Extensions icon on the toolbar and pin the Video Downloadhelper addon icon to the toolbar, so it's visible.
2. Left-click on the Video Downloadhelper icon on the toolbar 
3. Left-click on the 'settings' (gear) icon in the lower left corner of the screen.
4. Click on "More Settings" at the bottom of the menu that appears.
5. Scroll down to "Download directory" and click the "Change" link.
6. Browse to your desired download directory (or create a new one that you'd like to use), and click "select".
7. Verify new download location appears for "Download directory". 

Done...!

*NOTE:  if the CoApp is needed to download videos using Video Downloadhelper on your machine, a message will appear whenever you try to download a video telling you you need to install the CoApp, along with a link to download it from. Because of this, there's no need to include the installation of the CoApp as a part of the process of changing the download directory. If the Video Downloadhelper icon is showing on the toolbar, then everything you need to change the download directory using the steps above will be there.

**Reply by GrampaWildWilly** [reply]
Right.  I should have been a little clearer about how VDH uses the download directory.  That setting in the VDH settings is active only for the `Download` function.  To redirect a particular download to a different location, you have to go into the VDH settings & change the `Download directory` yourself before you start your `Download` function.  The function that is now called `Download` used to be called `Quick Download` in previous versions of VDH.

On the other hand, there is a function called `Download As`.  This function begins by prompting you for a directory into which to perform the download.  The prompting is done via a native operating system file selection dialog, also called a file picker.  This file picker also allows you to change the name of your target file.  But this feature is available only if you download via the CoApp.  The `Download As` function used to be called simply `Download` in previous versions of VDH.  The new `Download As` function does not refer to the `Download directory` . . . except for the very first invocation of the `Download As` function after you install VDH for the first time.  And when I say first time, I mean if you never had a previous version of VDH before.  Once you have done just one `Download As` function, VDH keeps the directory you used somewhere & uses that instead of the `Download directory`.  So there's 2 separate places where VDH remembers a previous download, one you can set via the settings, the other you can only influence by doing more `Download As` functions.  The old `Download` function used to update the `Download directory`.  The new `Download As` function does not do this.  Like I said, it only ever uses the `Download directory` setting once & after that it doesn't even refer to that setting, which means along the way that it does not update the setting the way it used to work in previous releases.  The only way to update the `Download directory` setting is to do it yourself in the VDH settings dialog.

**Reply by GrampaWildWilly** [reply]
If you are wondering where your next `Download As` function will go, you can get VDH to tell you.  Similarly, if you don't know where your most recent download of any kind went, you can get VDH to tell you this as well.

I have this in my VDH settings:

![image](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/4be64236-93d3-48bf-8107-910e1e8aff2e)

If I click that button labelled `Copy information to clipboard` there is a section in the result that looks like this:

```
    "download_directory": "M:\\VDH Testing",
    "last_download_directory": {
        "__serializer_tag": "option_some",
        "__serializer_value": {
            "__serializer_tag": "primitive",
            "__serializer_value": "M:\\Adult Time"
        }
```

The tag `"download_directory"` is the value of the setting named `Download directory` that you can see in the image above.  That is the directory VDH will use next time you execute a `Download` function.

The tag `"__serializer_value"` is the most recent directory I used on a `Download As` function.  It will be the directory first proposed to me next time I do a `Download As` function.  Except if you've never done a `Download As`.  In that case, the `"__serializer_value"` tag will be empty.

If you're wondering where your last download went to, or where your next one will go, click that copy to clipboard button in the VDH settings, then paste the results into your text editor.  Look for this section of what you pasted.  One or the other directory shown in this section of the pasted text will be where your download went or where your next download will go.

**Reply by Robubble** [reply]
Is there any place or way I could change it instead of in the settings? Currently (v9.5.0.2) in Firefox clicking that "change" button does nothing. The cursor changes on hover but that's all.

**Comment by NothernJock** [comment]
When you click the app icon on task bar, click the frame to add it to side bar there you can change the directory in settings then if you like close the sidebar again

**Reply by GrampaWildWilly** [reply]
You can change the directory in settings from the menu exactly the same as you can from the sidebar.  But like I explained, that setting applies only when you do a `Download` function.  You get to choose your own target directory on each `Download As` function.
